### DISCUSSION TITLE
async flush wait until flushed

### DISCUSSION THREAD

**Discussion by ghost** [gabime/spdlog#2084] [Q&A] [answered]
I'm wondering if the following approach is ok for waiting until a flush request on an async logger has been completed (this is in a situation where you guarantee there are no new messages coming in to be logged, otherwise may be waiting forever if queue is growing from new log requests):

```c++
// construct
spdlog::init_thread_pool(16384U, 6U);
auto spdlog_tp = spdlog::thread_pool(); 
auto async_logger = spdlog::create_async<sinks::basic_file_sink_mt>("logger_name", ...);

// ... do things

// flush when guaranteed no new logs coming in
async_logger.flush();
while(spdlog_tp->queue_size() != 0) { std::this_thread::sleep_for(Milliseconds(5)); }
```

**Comment by tt4g** [comment]
As long as you are using `spdlog::async_overflow_policy::block` (`spdlog::async_logger` default policy), calling `spdlog::shutdown()` is better than `while(spdlog_tp->queue_size() != 0) { std::this_thread::sleep_for(Milliseconds(5)); }`.

```diff
 async_logger.flush();
- while(spdlog_tp->queue_size() != 0) { std::this_thread::sleep_for(Milliseconds(5)); }
+ spdlog::shutdown();
```
Because:

In spdlog 1.9.2, `spdlog::details::thread_pool:` has a task queue, and executes tasks by pulling them from the queue in sequence.

https://github.com/gabime/spdlog/blob/eb3220622e73a4889eee355ffa37972b3cac3df5/include/spdlog/details/thread_pool-inl.h#L55-L86

`spdlog::async_logger::flush()` push flush task to queue of ``spdlog::details::thread_pool`.

https://github.com/gabime/spdlog/blob/eb3220622e73a4889eee355ffa37972b3cac3df5/include/spdlog/async_logger-inl.h#L40-L50

`spdlog::details::thread_pool::~thread_pool()` will push the `async_msg_type::terminate` task to the queue and wait for all tasks to finish.

https://github.com/gabime/spdlog/blob/eb3220622e73a4889eee355ffa37972b3cac3df5/include/spdlog/details/thread_pool-inl.h#L33-L53

`spdlog::shutdown()` will release `spdlog::details::thread_pool`.

https://github.com/gabime/spdlog/blob/a1d9f501e3863611824685426e3682cab1c8e016/include/spdlog/details/registry-inl.h#L253-L256


That is, as long as `spdlog::async_logger::flush()` is called before `spdlog::shutdown()`, the thread pool will be terminated after the flush operation.

**Reply by ghost** [reply]
Thanks for the response! What you've discussed works well if you intend for the logger to be shutdown after the flush. In my situation, I'd like to flush all messages before then continuing to allow further messages to be logged (e.g. I want to change the pattern being used, but first flush all logs from old pattern). 

For this use-case, do you have any suggested modifications for the original solution?

**Reply by tt4g** [reply]
If you can guarantee that no other log messages will be added to the thread pool queue, you can wait for the flush with the original solution.

**Reply by ghost** [reply]
Great! Could you add this as a separate answer so I can accept it?

**Comment by tt4g** [comment, marked as answer]
TO: https://github.com/gabime/spdlog/discussions/2084#discussioncomment-1265121

If you can guarantee that no other log messages will be added to the thread pool queue, you can wait for the flush with the original solution.
